src/librustc_mir/borrow_check/nll/constraint_set.rs

```diff
@@ -0,0 +1,33 @@
+use borrow_check::nll::region_infer::{ConstraintIndex, OutlivesConstraint};
+use rustc_data_structures::indexed_vec::IndexVec;
+use rustc_data_structures::fx::FxHashSet;
+use rustc::ty::RegionVid;
+
+#[derive(Clone, Default)]
+crate struct ConstraintSet {
+    constraints: IndexVec<ConstraintIndex, OutlivesConstraint>,
+    seen_constraints: FxHashSet<(RegionVid, RegionVid)>,
+}
+
+impl ConstraintSet {
+    pub fn push(&mut self, outlives_constraint: OutlivesConstraint) {
+        debug!("add_outlives({:?}: {:?} @ {:?}", outlives_constraint.sup, outlives_constraint.sub, outlives_constraint.point);
+        if outlives_constraint.sup == outlives_constraint.sub {
+            // 'a: 'a is pretty uninteresting
+            return;
+        }
+        if self.seen_constraints.insert(outlives_constraint.dedup_key()) {
+            self.constraints.push(outlives_constraint);
+        }
+    }
+
+    pub fn iner(&self) -> &IndexVec<ConstraintIndex, OutlivesConstraint> {
```

Comment:
Nit: "inner" -- but this method seems surprising, perhaps it'll disappear in future commits though =)

Reply:
I will fix the typo.

As to removing it in the future, iif that is what you want to get this merged. The alternative is to wrap all the iter methods that get used all over the place.